### PR TITLE
Escher interactive

### PR DESCRIPTION
**Pull request by axelvonkamp**

### CODE REVIEW

**Comment by Paulocracy** [comment]
So far, I didn't find any major issues with the Escher integration, thank you for this work! I just added a commit which makes the map menu a bit more consistent for the different map modes.

Up to now, I only found some small issues:
* When I create a new Escher map, the edit mode is active but the checkbox for the edit mode in the menu isn't. I didn't find a quick way to solve this, do you know one?
* Sometimes, after I calculate EFMs, a red text warning appears in the Escher box, but it disappears as soon as my mouse is entering the map.
* Should there be a warning that FVAs do not work with Escher maps?

**Comment by sthiele** [comment]
Just some quick feedback. I think it's not working for me under linux. The escher map is empty and when trying to drag some thing on the canvas get  this message:
`WebContentsAdapter::updateDragAction was not called within 3000 ms.`

On another note, I think its confusing if the map menu has different options depending on what kind of map one uses. I think it is best practice to keep the options visible and rather disable them or maybe move them to another place?
I think I also it would be nice if one could somehow distinguish escher maps from regular maps on the first glance.

I'm a teeny tiny bit afraid of the feature creep with javascript this would add  further complexity to the software while I would prefer to remove complexity. On the other hand :shrug: YOLO

File: cnapy/appdata.py [hunk 1]
```diff
@@ -21,7 +21,8 @@
 class AppData:
     ''' The application data '''
 
-    def __init__(self):
+    def __init__(self, qapp):
+        self.qapp = qapp
         self.version = "cnapy-1.1.1"
         self.format_version = 1
```
Comment (sthiele) [inline cnapy/appdata.py:27]: Would you also have to increase the format version?

**Comment by axelvonkamp** [comment]
> Just some quick feedback. I think it's not working for me under linux. The escher map is empty and when trying to drag some thing on the canvas get this message: `WebContentsAdapter::updateDragAction was not called within 3000 ms.`

Hmm, I don't think this is related to the OS. Currently there is a time delay of 1 s after escher_cnapy.html has issued the onload event and loading the actual map data. This is needed because even after onload Escher may not be fully operational because some Javascript is still being exectued. I have not yet found out how to find out if everything is operational, therefore the time delay. This may sometimes be too short and then result in an empty map. Can you try increasing the delay in the handle_onload_event function in escher_map_view.py and see if it works then? I will also try some things to get rid of the fixed delay.
Dragging reactions onto an Escher map is not supported, you need to use the edit mode in Escher for changing the map. I guess that dragging needs also to be explicitly rejected by the widget. 

> 
> On another note, I think its confusing if the map menu has different options depending on what kind of map one uses. I think it is best practice to keep the options visible and rather disable them or maybe move them to another place? I think I also it would be nice if one could somehow distinguish escher maps from regular maps on the first glance.

I don't know if lots of disabled options look better, I think this way it is cleaner. Escher maps look quite different from regular maps, I think they can be easily distinguished.

> 
> I'm a teeny tiny bit afraid of the feature creep with javascript this would add further complexity to the software while I would prefer to remove complexity. On the other hand shrug YOLO

Oh, it definitely adds complexity and because all communication with the Javascript side is asynchronous getting it to work has been hell xD. But also a good learning experience :-)

File: cnapy/appdata.py (hunk 1 above)
Comment (axelvonkamp) [reply cnapy/appdata.py:27]: Good point, will set
self.format_version = 2

**Comment by axelvonkamp** [comment]
> So far, I didn't find any major issues with the Escher integration, thank you for this work! I just added a commit which makes the map menu a bit more consistent for the different map modes.
> 
> Up to now, I only found some small issues:
> 
>     * When I create a new Escher map, the edit mode is active but the checkbox for the edit mode in the menu isn't. I didn't find a quick way to solve this, do you know one?

Thanks, will fix this.

> 
>     * Sometimes, after I calculate EFMs, a red text warning appears in the Escher box, but it disappears as soon as my mouse is entering the map.

Do you mean the hint that the visualization scheme can be changed in the settings or some other warning?

>     * Should there be a warning that FVAs do not work with Escher maps?

I don't know. FVA is currently just not visualized. I will check if it is possible to display a flux range instead of a single value.

**Comment by Paulocracy** [comment]
> Do you mean the hint that the visualization scheme can be changed in the settings or some other warning?

Ah yes this is the warning I thought it is Escher-related but if it's from CNApy it's fine.

> I don't know. FVA is currently just not visualized. I will check if it is possible to display a flux range instead of a single value.

I think that the user should be at least aware of it, displaying a range would be great as EFMler wasn't able to do it so that it relies on colors (which is also not optimal).

I also agree with Sven that this JavaScript implementation is very complex, but at least we all have some experience with JavaScript integrations :3 and Escher is just widely requested as it is perceived as a kind of map standard.

And regarding the UX problems with the menus, maybe changing the "Map" tab caption to "Map (Escher mode)" or "Map (normal mode)", depending on the current map, would help?

**Comment by sthiele** [comment]
Hmm still not working for me. I don't get the timeout error anymore but I still see nothing, Also trying to access Escher settings has no effect for me. If I create a new Escher map the canvas just flashes twice and then stays empty.
Maybe I'm missing some dependency? Most concerning is that there is no error message which makes it hard to debug,

**Comment by axelvonkamp** [comment]
> Hmm still not working for me. I don't get the timeout error anymore but I still see nothing, Also trying to access Escher settings has no effect for me. If I create a new Escher map the canvas just flashes twice and then stays empty. Maybe I'm missing some dependency? Most concerning is that there is no error message which makes it hard to debug,

I have made some changes to the setup procedure. Although I have no real idea why it was not working for you, maybe it is now? If not, which Qt version do you use? Can you send me screenshots of the inoperative Escher tabs?

**Review by Paulocracy — APPROVED** [review]
Thank you for this huge commit :D Although @sthiele still has (non-reproducible) major issues with the Escher maps, I decided to approve it since the problems only affect the Escher maps, so that the new release notes should add a warning and a message to call for bug reports :-|

**Pull request merged**